pkgpanda/util.py

```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
     return session
 
 
+def stream_remote_file_with_retries(out_filename, url, retries=4):
+    def cleanup():
+        try:
+            os.remove(out_filename)
+        except Exception:
+            pass
+
+    while True:
+        with open(out_filename, "w+b") as f:
+            r = get_requests_retry_session().get(url, stream=True)
+            if r.status_code == 301:
+                raise Exception("got a 301")
+            r.raise_for_status()
+
+            for chunk in r.iter_content(chunk_size=4096):
+                f.write(chunk)
+            f.flush()
+
+            content_length = int(r.headers['content-length'])
+            final_size = os.fstat(f.fileno()).st_size
+            if final_size == content_length:
+                return r
+            else:
+                f.close()
+                cleanup()
+                retries -= 1
```

Comment:
There's a very lovely little python library called `retrying` (https://pypi.org/project/retrying/) that let's you annotate a function (even an inline function) with retry logic of this sort.) We should use that instead of rolling our own.

For example:
https://github.com/dcos/dcos/blob/5391993fa0ca8457dae4d2126d2c0c515dd8264a/packages/dcos-integration-test/extra/util/delete_ec2_volume.py#L46
https://github.com/dcos/dcos/blob/57350f12eb7c1ed82eeb7e8944c72776b936160d/packages/dcos-integration-test/extra/test_networking.py#L60-L62

I like declaring little inner functions in my larger function and annotating the inner ones with retrying logic.

Reply:
I see we had the same thought! https://github.com/dcos/dcos/pull/4549/files#r258438916

Reply:
Switched to `retrying` — it's lovely!